Make multi-arg options optional

Rishi-0007 opened the pull request:
Fixes #331 Updates the CLI handler to make multi-argument options optional. This change prevents single-value options, like `-k` or `--kit`, from capturing unintended positional arguments. Only one value is assigned to options, leaving other arguments as positional. Additionally, a unit test has been added to verify this behavior.

#### Test:
```typescript
import { Command } from '../src/cli/utils/cli-handler'

describe('CLI Handler Tests', () => {
	it('should assign only a single argument to -k option', () => {
		const command = new Command('test-command')
		command.option('-k', '--kit', 'Test option')

		const args = ['-k', 'activity', 'myactivity']
		const parsedOptions = command['parseOptions'](args)

		expect(parsedOptions.kit).toBe('activity')
		expect(args).toContain('myactivity')
	})
})

```

#### Test result:
```bash
npx jest cli-handler.test.ts
 PASS  __tests__/cli-handler.test.ts
  CLI Handler Tests
    ✓ should assign only a single argument to -k option (2 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        2.249 s
Ran all test suites matching /cli-handler.test.ts/i.

```

Pkmmte commented:
**CLA Assistant Lite bot:** <br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/Wave-Play/robo.js/blob/main/CLA.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   <sub>You can retrigger this bot by commenting **recheck** in this Pull Request</sub>

Rishi-0007 commented:
I have read the CLA Document and I hereby sign the CLA

Pkmmte commented:
Thank you for this! Can you please verify that multi-arg options are still possible? If so, how does the API distinguish between them?

For example, a `--plugins` option should still be able to accept multiple values. (e.g. "--plugins @robojs/ai @robojs/sync")

Rishi-0007 commented:
> Thank you for this! Can you please verify that multi-arg options are still possible? If so, how does the API distinguish between them?
> 
> For example, a `--plugins` option should still be able to accept multiple values. (e.g. "--plugins @robojs/ai @robojs/sync")

Hey @Pkmmte , I've made the changes based on your feedback. Please check it and let me know if this approach is good or any further changes are required.

Pkmmte commented:
Hi, we've marked this submission as accepted for Hacktoberfest purposes because there's some improvements in here we like, but overall we've found some regressions.

### For example, positional arguments no longer work.

```bash
npx robo build src/commands/ping.ts -v
```

For some reason, it treats options (`-v`) as positional arguments themselves.

```
info  - Building Robo...
node:internal/fs/promises:1037
  const result = await PromisePrototypeThen(
                 ^

Error: ENOENT: no such file or directory, stat '/Users/pkmmte/demo/cli-update-1/-v'
    at async Object.stat (node:internal/fs/promises:1037:18)
    at async traverse (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/compiler.js:59:18)
    at async Object.buildCode (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/compiler.js:133:3)
    at async Command.buildAction [as _handler] (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/commands/build/index.js:63:23)
    at async Command.processSubCommand (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/cli-handler.js:255:5)
    at async Command.processSubCommand (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/cli-handler.js:245:9) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/pkmmte/demo/cli-update-1/-v'
}

Node.js v22.2.0
```

(ENOENT in this case because `build` looks for positional arguments in the file system)

### Changes expected behavior

The type signature and behavior for multiple options (enabled by default) is also changed. Previously, options with spaces would still yield a single string, but it looks like this now changes to a string array. 

While I agree that an array would be more appropriate, we like to avoid changing behavior like this unless it's a major version release. It is likely to break existing code reliant on options always being a single string. In fact, our own `create-robo` package would break from this change because of how it uses `--plugins`.

As stated earlier, we will be accepting this PR for Hacktoberfest purposes, but may delay in merging due to the above issues.

Pkmmte reviewed (changes requested):
Hi, we've marked this submission as accepted for Hacktoberfest purposes because there's some improvements in here we like, but overall we've found some regressions.

### For example, positional arguments no longer work.

```bash
npx robo build src/commands/ping.ts -v
```

For some reason, it treats options (`-v`) as positional arguments themselves.

```
info  - Building Robo...
node:internal/fs/promises:1037
  const result = await PromisePrototypeThen(
                 ^

Error: ENOENT: no such file or directory, stat '/Users/pkmmte/demo/cli-update-1/-v'
    at async Object.stat (node:internal/fs/promises:1037:18)
    at async traverse (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/compiler.js:59:18)
    at async Object.buildCode (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/compiler.js:133:3)
    at async Command.buildAction [as _handler] (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/commands/build/index.js:63:23)
    at async Command.processSubCommand (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/cli-handler.js:255:5)
    at async Command.processSubCommand (file:///Users/pkmmte/demo/cli-update-1/node_modules/robo.js/dist/cli/utils/cli-handler.js:245:9) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: '/Users/pkmmte/demo/cli-update-1/-v'
}

Node.js v22.2.0
```

(ENOENT in this case because `build` looks for positional arguments in the file system)

### Changes expected behavior

The type signature and behavior for multiple options (enabled by default) is also changed. Previously, options with spaces would still yield a single string, but it looks like this now changes to a string array. 

While I agree that an array would be more appropriate, we like to avoid changing behavior like this unless it's a major version release. It is likely to break existing code reliant on options always being a single string. In fact, our own `create-robo` package would break from this change because of how it uses `--plugins`.

As stated earlier, we will be accepting this PR for Hacktoberfest purposes, but may delay in merging due to the above issues.

Rishi-0007 commented:
Hi @Pkmmte ,

I've updated `cli-handler.ts` and the tests as per your feedback. The option parsing now correctly handles options after positional arguments, and options with spaces return a single string by default to maintain backward compatibility.

All tests are passing:

```
PASS  __tests__/cli-handler.test.ts
  CLI Handler Tests
    ✓ should correctly parse options after positional arguments
    ✓ should return single string for options with spaces by default
    ✓ should return array for options when returnArray is true
```

However, I'm still encountering the `ENOENT` error when running:

```bash
npx robo build src/commands/ping.ts -v
```

It seems the `-v` option is still being treated as a positional argument, causing the build command to look for a file named `-v`.

Could you please provide more details or guidance on resolving this issue? I want to ensure the PR doesn't introduce any regressions and is ready for merging.

Thank you for your time!

Rishi-0007 commented:
Hi @Pkmmte,

Thank you for your feedback and for accepting the PR for Hacktoberfest. I’ve been working on addressing the issues mentioned, specifically ensuring options like `-v` after positional arguments are correctly parsed and not treated as file paths.

Here’s what I’ve tried so far:

1. **Adjusted `parseOptions` in `cli-handler.ts`**: Refined the parsing logic to separate options from positional arguments and ensure options with spaces return single strings by default.
2. **Verified Argument Parsing with Tests**: Added and confirmed tests for scenarios with options appearing after positional arguments, ensuring options are recognized correctly.
3. **Encountering Persistent `ENOENT` Error**: When running the command:

   ```bash
   npx robo build src/cli/commands/add.ts -v
   ```

   the `-v` flag continues to be interpreted as a positional argument, causing an attempt to locate a file at `-v`, and resulting in the `ENOENT` error. This behavior appears on both the main branch and my modified branch, suggesting it may be due to the overall handling of positional arguments in the `build` command.

Would you be able to provide additional guidance on addressing this issue? Specifically, are there any expected parsing behaviors or modifications within `cli-handler.ts` that would prevent options like `-v` from being treated as positional arguments in the context of the `build` command?

Thank you for your time and assistance.

Nazeofel commented:
Hello ! thanks for your contribution, I have just tested it out, and it seems to be working just fine for me, are you sure to be testing out the right branch ?

Rishi-0007 commented:
> Hello ! thanks for your contribution, I have just tested it out, and it seems to be working just fine for me, are you sure to be testing out the right branch ?

Hi @Nazeofel, Thanks for testing it out. 
Is this the expected output or am I doing something wrong?

```bash
npx robo build src/cli/commands/add.ts 
info  - Building Robo...
node:internal/modules/esm/resolve:839
  throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base), null);
        ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'robo.js' imported from /home/rishi/code/robo.js/p
ackages/robo/.robo/build/commands/help.js                                                               at packageResolve (node:internal/modules/esm/resolve:839:9)
    at moduleResolve (node:internal/modules/esm/resolve:908:18)
    at defaultResolve (node:internal/modules/esm/resolve:1038:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:557:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:525:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:246:38)
    at ModuleJob._link (node:internal/modules/esm/module_job:126:49) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v22.9.0

```

Rishi-0007 commented:
Hey @Pkmmte,

I’ve tested the issue on the main branch as well and observed the following outputs in two cases:

- **Without `-v`**:
  ```bash
  npx robo build src/cli/commands/add.ts
  info  - Building Robo...

  Type                Name            Description
  ───────────────────────────────────────────────
  Command            Δ /help            Displays a list of commands.
  Event              Δ interactionCreate
  Event              Δ ready  

  Δ = Automatically generated

  Robo size: 836.59 kB
  Built in 109ms

  discord:error - DISCORD_TOKEN or DISCORD_CLIENT_ID not found in environment variables
  ```

- **With `-v`**:
  ```bash
  npx robo build src/cli/commands/add.ts -v
  info  - Building Robo...
  node:internal/fs/promises:1032
    const result = await PromisePrototypeThen(
                   ^

  Error: ENOENT: no such file or directory, stat '/home/rishi/code/robo.js/packages/robo/-v'
      at async Object.stat (node:internal/fs/promises:1032:18)
      at async traverse (file:///home/rishi/code/robo.js/packages/robo/dist/cli/utils/compiler.js:59:1
  8)                                                                                                      at async Object.buildCode (file:///home/rishi/code/robo.js/packages/robo/dist/cli/utils/compiler
  .js:133:3)                                                                                              at async Command.buildAction [as _handler] (file:///home/rishi/code/robo.js/packages/robo/dist/c
  li/commands/build/index.js:63:23)                                                                       at async Command.processSubCommand (file:///home/rishi/code/robo.js/packages/robo/dist/cli/utils
  /cli-handler.js:255:5)                                                                                  at async Command.processSubCommand (file:///home/rishi/code/robo.js/packages/robo/dist/cli/utils
  /cli-handler.js:245:9) {                                                                              errno: -2,
    code: 'ENOENT',
    syscall: 'stat',
    path: '/home/rishi/code/robo.js/packages/robo/-v'
  }

  Node.js v22.9.0
  ```

This `ENOENT` error seems unrelated to my changes since it’s happening on the main branch as well. Could you confirm if I’m missing something or if there’s another approach I should take? Your guidance would be greatly appreciated.

Thank you!

Nazeofel commented:
Hey ! it should work just fine on the main branch and latest version of Robo, just to confirm somethin, which OS are you running ?

Rishi-0007 commented:
> Hey ! it should work just fine on the main branch and latest version of Robo, just to confirm somethin, which OS are you running ?

Ubuntu noble 24.04 x86_64

Nazeofel commented:
Hey sorry to get back to you so late, we cannot reproduce this issue at all on our environment with the main branch, tested on MacOS, Windows, and Linux too. would you be able to share your all Linux configuration where you are running the command please.
